### ISSUE TITLE
Architectural fragility: trained predator can't survive inference-time role_prefix length changes

### ISSUE THREAD

**Issue by dgonier** [dgonier/consciousness-ecology-codd#12]
## Status
PROPOSED

## Motivation
While running Experiment A in #11 (FinCoT structured-CoT prompting probe), discovered that the trained predator policy **cannot survive an inference-time role_prefix swap** of meaningful length. Replacing the 604-char training prompt with a 1201-char structured prompt produced pure gibberish output (`PREDi p r e d Ie THER E R. SPOILER: P REDI CERE DETER...`), not the expected structured CoT.

This is a real architectural fragility we didn't know about until #11 surfaced it.

## Diagnosis

The cross-attention Q is built (post #6 fix) as `role_q + prey_q` where `role_q = self.role_prefix.mean(dim=0)`. Doubling the role prefix length:

1. Shifts `role_prefix.mean(dim=0)` to a different point in 2560-dim hidden space
2. Trained Channel `W_K` weights expect the original Q geometry
3. Q · K alignment collapses
4. Channel output → garbage
5. Predator decode → degenerate token sequence

Verified via `scripts/diagnostics/exp_A_inspect.py`: running on `ipo_seed8_best.pt` with max_new_tokens=400 shows character-level decay across all token positions.

## What this implies
- **Role prefixes are effectively frozen at training time.** Inference-time prompt changes don't compose with trained Channels.
- **All future "let's just try a different prompt" experiments need to retrain**, not just patch and run.
- **The paper claim space narrows**: we can't claim "this architecture handles structured prompting" because it doesn't unless we retrain for it.

## Why this isn't trivial to fix
The natural fix is to make the predator's Q construction *robust* to prompt-length shifts:
- **Option 1**: Use the role_prefix's *last token* hidden state, not the mean. Last-token is more stable across prompt lengths since the trained policy always sees the prompt-ending position.
- **Option 2**: Normalize role_q to a fixed norm before adding prey_q. This decouples geometry from length.
- **Option 3**: Use a learned projection from role_prefix → fixed-dim Q anchor that's prompt-length-invariant by design.
- **Option 4**: Keep the current Q construction but train with prompt-length augmentation so the policy is robust to it.

Option 2 is the smallest diff and probably enough. Would need to retrain to verify.

## Why this matters for #8 (and beyond)
The whole point of #11 was a CHEAP test of structured-CoT prompting on direction collapse. We can't do cheap tests because the architecture is rigid. Every new prompting hypothesis is now a multi-hour retrain, not a 25-min eval.

This also affects:
- Director Nodes (#2) — if the apex prompts the predator with directives, those need to be in the trained distribution
- Per-species pretraining (#10) — each species needs prompt-length consistency between CPT, SFT, IPO, and inference
- Future eval tooling — can't sweep prompts for robustness analysis

## Context pointers
- Files: `trophic/agents/predator.py:hunt_and_predict` — Q construction
- Files: `trophic/training/sft.py:_channel_output_for` and the IPO/GRPO/reward_rl mirrors — same Q construction at training time
- Logs: `logs/exp_A_fincot.log`, output of `scripts/diagnostics/exp_A_inspect.py`
- Related issues: #6 (the original role_prefix-based Q construction this fragility lives in), #11 (the experiment that surfaced this), #10 (per-species CPT — which would all share this fragility unless fixed)

## Dependencies
- Blocked by: nothing
- Blocks: #11 (declined for now; could be reopened if this is fixed), prompt-side iterations on #2, #10
- Independent of: #1, #3

## Trigger conditions
Start when (a) #10's first species CPT run is being designed — perfect time to fold in the Q-robustness fix so we don't bake the fragility into a per-species checkpoint, OR (b) someone else proposes a prompt-side experiment and we want it to actually be testable.

## Decision criteria
- Q-construction fix lands (one of the four options)
- Role prefix length can vary at inference time without breaking the trained policy (test: same `ipo_seed9_best.pt` with two prompt variants of different lengths produces parseable output for both)
- Existing 83 tests still pass

## Open questions
- Is this only about role_prefix length, or does CONTENT shift also matter? Likely both. Test: same length but different blueprint → does the policy still work?
- Does this fragility affect herbivores too, or just predator? The same Q-construction pattern is in herbivore.py. Probably affects both.

## Effort estimate
- Option 2 (norm normalization): ~30 min code change + ~3 hr retrain to verify
- Option 1 (last-token): ~1 hr code change + retrain + sanity test
- Option 4 (training-time augmentation): ~1 day to design + retrain

## Out of scope
- Fixing prompt-content-shift fragility (separate issue if it exists)
- LoRA / adapter approaches to making prompts swappable (that's a different architecture)

**Comment by dgonier** [comment]
## Partial fix landed — but the deeper fragility is in Qwen's next-token decoder, not our Channels

### What landed

`trophic/agents/base.py::normalize_role_q` — pools `role_prefix.mean(dim=0)` then normalizes to `ROLE_Q_REF_NORM=1.0`. Applied at all 6 Q-construction sites:
- `agents/predator.py`
- `agents/herbivore.py`
- `training/sft.py` (×2 sites — `_channel_output_for` and the producer-trough skip path)
- `training/ipo.py`, `grpo.py`, `reward_rl.py`

6 new unit tests in `tests/test_role_q_normalization.py` lock in length-invariance + dtype preservation + pathological-input handling. **89/89 tests pass** (was 83 + 6 new).

### What this fix actually solves

The cross-attention Channel layer's Q geometry is now length-invariant. That was step 1 of the diagnosis — and the right fix at the level it operates on. Anyone retraining with a different prompt now gets stable Q geometry across runs.

### What this fix does NOT solve

Re-ran `scripts/diagnostics/exp_A_inspect.py` (which originally produced gibberish). **Output is bit-identical to before the fix — still gibberish.** The Q-geometry was not actually the load-bearing failure mode.

The real fragility is in `model_host.py::forward_with_prefix`:

```python
inputs_embeds = [role_prefix, channel_output, query_text_embeds]
```

The ENTIRE role_prefix (full sequence, not pooled, not normalized) is fed directly into Qwen as the prefix tokens. **Qwen's next-token decoder produces gibberish when given an out-of-distribution prefix** — pretraining never saw "STEP 1 — IDENTIFY..." as a prefix to financial XML output, so the language-model head fails.

This isn't a fragility in our architecture per se. It's a fundamental property of **prefix-injection-based prompting**: the only way to change the prompt is to retrain the model with the new prompt in its SFT/IPO distribution.

### Reframe

The "cheap test of FinCoT prompting" idea was always going to fail — not because of our architecture, but because LLMs in general can't survive arbitrary prefix swaps without retraining. This is a property of the host model, not of the trophic stack.

The Q-normalization fix is still correct and useful:
- It makes the Channel layer length-invariant for *future* retraining at different prompt lengths
- It cleanly separates Q-geometry from prefix-content variation
- It defensively prevents the smaller-magnitude failure mode if/when it surfaces independently

But for the original goal (test FinCoT prompting cheaply on `ipo_seed8_best.pt`), the answer is: **not possible without retraining**. Routing to #10 (per-species CPT) where the new prompt format gets baked into training.

### Updated status

- Q-normalization fix: **DONE**, tested, 6 new tests
- Prompt-swap-without-retrain capability: **NOT ACHIEVABLE** without changing how role_prefix is consumed downstream (would require either prefix tuning, prompt-tuning embeddings, or a learned prefix adapter — all bigger than this fix)
- Issue closes as **status:done** for the surgical Q fix; the broader "make prompts swappable at inference time" stays open as a future direction (probably its own issue if/when needed)

### Honest paper-claim implication

Future work on this stack must treat role prefixes as fixed at training time. Each new prompt structure = one CPT/SFT/IPO cycle. There is no shortcut.
